File: logibra.py
import sys
import re

# --- 1. Data Structures (AST) ---

class Node:
    pass

class Unit(Node):
    def __init__(self, modifiers):
        self.modifiers = modifiers  # List of modifiers
        self.anchored = '@' in modifiers
        self.prime_count = modifiers.count("'")

    def __repr__(self):
        mods = ("@" if self.anchored else "") + ("'" * self.prime_count)
        return f"*{mods}"

    def __eq__(self, other):
        return (isinstance(other, Unit) and
                self.anchored == other.anchored and
                self.prime_count == other.prime_count)

class Flow(Node):
    def __init__(self, unit, polarity):
        self.unit = unit
        self.polarity = polarity  # String: /+, \+, /-, \-

    def __repr__(self):
        return f"{self.unit} -> {self.polarity}"

class Relation(Node):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def __repr__(self):
        return f"({self.left} & {self.right})"

class Resolution(Node):
    def __repr__(self):
        return "**"

# --- 2. Reduction Logic (The Engine) ---

def are_inverses(pol_a, pol_b):
    """
    Quadrant-Dominant Logic:
    Total Inversion requires Opposite Slope AND Opposite Sign.
    Pairs: (/+ vs \-) and (\+ vs /-)
    """
    pairs = [
        ('/+', '\-'), ('\-', '/+'),
        ('\+', '/-'), ('/-', '\+')
    ]
    return (pol_a, pol_b) in pairs

def reduce(node):
    """
    Recursive reduction engine.
    Returns: The simplified AST.
    """
    # Base cases
    if isinstance(node, (Unit, Resolution)):
        return node

    # Reduce Flow (verify unit is resolved)
    if isinstance(node, Flow):
        reduced_unit = reduce(node.unit)
        # If the unit disappeared or became Resolution (impossible in valid grammar but safe to check)
        if isinstance(reduced_unit, Resolution):
            return Resolution()
        return Flow(reduced_unit, node.polarity)

    # Reduce Relation (The Core Logic)
    if isinstance(node, Relation):
        left = reduce(node.left)
        right = reduce(node.right)

        # Rule: ** absorbs everything (Equality is terminal)
        if isinstance(left, Resolution) or isinstance(right, Resolution):
            return Resolution()

        # Rule: Cancellation (Total Inversion)
        # Structure must be: Flow(U1, P1) & Flow(U2, P2)
        if isinstance(left, Flow) and isinstance(right, Flow):
            # 1. Units must be identical (including modifiers)
            if left.unit == right.unit:
                # 2. Polarities must be total inverses
                if are_inverses(left.polarity, right.polarity):
                    return Resolution()

        # If no reduction occurred, return the relation
        return Relation(left, right)

    return node

# --- 3. Parser (Recursive Descent) ---

class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def peek(self):
        return self.tokens[self.pos] if self.pos < len(self.tokens) else None

    def consume(self, expected=None):
        token = self.peek()
        if expected and token != expected:
            raise ValueError(f"Expected {expected}, got {token}")
        self.pos += 1
        return token

    def parse_expression(self):
        # Check for standalone Resolution "**"
        if self.peek() == '**':
            self.consume()
            return Resolution()

        left = self.parse_term()
        while self.peek() == '&':
            self.consume('&')
            right = self.parse_term()
            left = Relation(left, right)
        return left

    def parse_term(self):
        if self.peek() == '(':
            self.consume('(')
            expr = self.parse_expression()
            self.consume(')')
            return expr

        # Must be an Atom
        unit = self.parse_atom()

        # Check for Flow
        if self.peek() == '->':
            self.consume('->')
            pol = self.consume()
            if pol not in ['/+', '\+', '/-', '\-']:
                 raise ValueError(f"Invalid polarity: {pol}")
            return Flow(unit, pol)

        return unit

    def parse_atom(self):
        self.consume('*')
        modifiers = []
        while self.peek() in ['@', "'", ',']:
            mod = self.consume()
            if mod != ',': # Ignore commas
                modifiers.append(mod)
        return Unit(modifiers)

def tokenize(text):
    # Order matters: ** before *, complex polarity before simple symbols
    token_pattern = r"(\*\*|->|/\+|\\+|/-|\\-|&|\(|\)|\*|@|'|,)"
    tokens = [t for t in re.split(token_pattern, text.replace(" ", "")) if t]
    return tokens

# --- 4. Main Driver ---

def run_logibra(text):
    try:
        tokens = tokenize(text)
        parser = Parser(tokens)
        ast = parser.parse_expression()
        result = reduce(ast)
        print(f"Input:  {text}")
        print(f"Parsed: {ast}")
        print(f"Result: {result}")
        print("-" * 20)
        return result
    except Exception as e:
        print(f"Error parsing '{text}': {e}")
        return None

if __name__ == "__main__":
    # Standard Examples
    print("--- Logibra Standard Execution ---\n")
    run_logibra("(* -> /+) & (* -> \-)")
    run_logibra("*@ -> \+")
    run_logibra("(* -> /+) & (* -> /+)")

File: Logibra.hs
module Logibra where

-- 1. Data Types (The Universe) --

data Modifier = Anchor | Prime
    deriving (Show, Eq)

data Polarity = P_UpRight   -- /+ (Q1)
              | P_DownRight -- \+ (Q4)
              | P_DownLeft  -- /- (Q3)
              | P_UpLeft    -- \- (Q2)
    deriving (Show, Eq)

data Expr = Unit [Modifier]          -- * with modifiers
          | Flow Expr Polarity       -- * -> Pol
          | Relation Expr Expr       -- A & B
          | Resolution               -- **
    deriving (Eq)

-- Custom Show for cleaner output
instance Show Expr where
    show (Unit mods) = "*" ++ concatMap showMod mods
      where showMod Anchor = "@"
            showMod Prime  = "'"
    show (Flow e p) = show e ++ " -> " ++ showPol p
      where showPol P_UpRight   = "/+"
            showPol P_DownRight = "\\+"
            showPol P_DownLeft  = "/-"
            showPol P_UpLeft    = "\\-"
    show (Relation a b) = "(" ++ show a ++ " & " ++ show b ++ ")"
    show Resolution = "**"

-- 2. The Logic (Reduction Engine) --

-- Quadrant-Dominant Inversion Logic
-- Only opposite slope AND opposite sign cancel.
areInverses :: Polarity -> Polarity -> Bool
areInverses P_UpRight P_UpLeft    = True  -- /+ cancels \-
areInverses P_UpLeft  P_UpRight   = True
areInverses P_DownRight P_DownLeft = True -- \+ cancels /-
areInverses P_DownLeft P_DownRight = True
areInverses _ _                   = False

reduce :: Expr -> Expr
reduce Resolution = Resolution
reduce (Unit m) = Unit m
reduce (Flow e p) =
    let re = reduce e in
    case re of
        Resolution -> Resolution -- Flow cannot contain resolved state
        _          -> Flow re p

reduce (Relation a b) =
    let ra = reduce a
        rb = reduce b
    in case (ra, rb) of
        (Resolution, _) -> Resolution -- Absorption rule
        (_, Resolution) -> Resolution

        -- The Core Cancellation Rule
        (Flow u1 p1, Flow u2 p2) ->
            if u1 == u2 && areInverses p1 p2
            then Resolution
            else Relation ra rb

        -- Default: keep relation
        _ -> Relation ra rb

-- 3. Simple Test Runner --

main :: IO ()
main = do
    putStrLn "--- Logibra Haskell Reducer ---"

    -- Test 1: Cancellation (* -> /+) & (* -> \-)
    let t1 = Relation (Flow (Unit []) P_UpRight) (Flow (Unit []) P_UpLeft)
    printTestCase "Standard Opposition" t1

    -- Test 2: Conflict (* -> /+) & (* -> /+)
    let t2 = Relation (Flow (Unit []) P_UpRight) (Flow (Unit []) P_UpRight)
    printTestCase "Conflict (Same Pol)" t2

    -- Test 3: Anchored Flow (*@ -> \+)
    let t3 = Flow (Unit [Anchor]) P_DownRight
    printTestCase "Anchored Flow" t3

printTestCase :: String -> Expr -> IO ()
printTestCase name expr = do
    putStrLn $ "Test: " ++ name
    putStrLn $ "Input:  " ++ show expr
    putStrLn $ "Output: " ++ show (reduce expr)
    putStrLn "----------------"
